tso-server/src/main/java/org/apache/omid/tso/PersistenceProcessorImpl.java

```diff
@@ -204,15 +204,13 @@ private void maybeFlushBatch() {
     synchronized private void flush() {
         lastFlush.set(System.nanoTime());
 
-        boolean areWeStillMaster = true;
         if (!leaseManager.stillInLeasePeriod()) {
-            // The master TSO replica has changed, so we must inform the
-            // clients about it when sending the replies and avoid flushing
-            // the current batch of TXs
-            areWeStillMaster = false;
-            // We need also to clear the data in the buffer
-            writer.clearWriteBuffer();
-            LOG.trace("Replica {} lost mastership before flushing data", tsoHostAndPort);
+            if (batch.getNumEvents() > 0) {
+                // The master TSO replica has changed, so we must commit suicide
+                // We need also to clear the data in the buffer
+                writer.clearWriteBuffer();
+                panicker.panic("Replica " + tsoHostAndPort + " lost mastership before flushing data. Committing suicide");
```

Comment:
Done in rebased commit like this:
    synchronized private void flush() {

        if (batch.getNumEvents() > 0) {
            lastFlush.set(System.nanoTime());
            commitSuicideIfNotMaster();
            try {
                writer.flush();
                batchSizeHistogram.update(batch.getNumEvents());
            } catch (IOException e) {
                panicker.panic("Error persisting commit batch", e);
            }
            commitSuicideIfNotMaster(); // TODO Here, we can return the client responses before committing suicide
            flushTimer.update((System.nanoTime() - lastFlush.get()));
            batch.sendRepliesAndReset(reply, retryProc);
        }

    }

    private void commitSuicideIfNotMaster() {
        if (!leaseManager.stillInLeasePeriod()) {
            panicker.panic("Replica " + tsoHostAndPort + " lost mastership whilst flushing data. Committing suicide");
        }
    }

Some tests have been modified